OpenRA.Mods.Common/Projectiles/AreaBeam.cs

```diff
@@ -138,7 +138,8 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
 			target += dir * info.BeyondTargetRange.Length / 1024;
 
 			length = Math.Max((target - headPos).Length / speed.Length, 1);
-			weaponRange = new WDist(Util.ApplyPercentageModifiers(args.Weapon.Range.Length, args.RangeModifiers));
+			weaponRange = new WDist(Util.ApplyPercentageModifiers(args.Weapon.Range.Length, args.PercentRangeModifiers));
+			weaponRange = Util.ApplyFlatDistanceModifiers(weaponRange, args.FlatRangeModifiers);
```

Comment:
Don't know how often this is called. But better would be to have 2 if blocks. To only calculate flat range if set.

Similar in other lcoations below.

Reply:
Would optionally suggest to remove the `Flat` prefix or rename it to `Absolute`.

Reply:
> Don't know how often this is called. But better would be to have 2 if blocks. To only calculate flat range if set.
> 
> Similar in other lcoations below.

Inside ApplyFlatDistanceModifiers it checks if any modifiers are set. Think that'd be preferable to duplicating the check outside every time?

> Would optionally suggest to remove the Flat prefix or rename it to Absolute.

Happy to change it if others think so too, but personally I see "flat" as the counterpart to "percentage" (or "proportional"), whereas "absolute" would be the counterpart to "relative" (which I also considered, but felt that it begs the question "relative to what?" whereas percentage is a bit clearer).

Reply:
> Inside ApplyFlatDistanceModifiers it checks if any modifiers are set. Think that'd be preferable to duplicating the check outside every time?

- would percentage be set more often in future than flat? perhaps you can do the same check in the apply percentage function.


the code is a hot path. the apply percentage function even has a comment on performance. 

- it would save looking up the modifiers each time i.e. `ai.TraitInfos<IFlatRangeModifierInfo>().Select(m => m.GetRangeModifierDefault())` 
- you now check twice if modifiers are set, once via length, once via Any.
- does it make sense to have multiple fixed modifiers? 
- couldn't you sum them just as well before - rather than sum them each time, i assume? the LINQ  `Sum` is 'expensive'.
- both apply functions might be faster if the function argument was an array and not an enumerable. not sure. perhaps worth testing.

let's see what the others say.

Reply:
A percentage *is* relative, so absolute would indeed be appropriate here.

Reply:
> A percentage _is_ relative, so absolute would indeed be appropriate here.

It'd be appropriate sure, just personally I wouldn't have Absolute & Percent, I'd have either Absolute & Relative, or Flat & Percent. Happy to rename Flat to Absolute if that's what you'd prefer though.

>     * would percentage be set more often in future than flat? perhaps you can do the same check in the apply percentage function.

Not sure how beneficial it'd be there, as it just iterates over the percentages; would similarly iterating over the flat modifiers (instead of using Sum()) be better and remove the need to do Any()?

>     * it would save looking up the modifiers each time i.e. `ai.TraitInfos<IFlatRangeModifierInfo>().Select(m => m.GetRangeModifierDefault())`

Hmm not sure what you mean here. Afaik ModifiedRange is only used for rendering range circles.

>     * you now check twice if modifiers are set, once via length, once via Any.

I think the only place I used .Length is to get that property from the WDist of each modifier, so it's not using it like Any(), if that's what you meant?

>     * does it make sense to have multiple fixed modifiers?

Yea, you can have mutliple buffs or debuffs that affect the same attributes.

>     * couldn't you sum them just as well before - rather than sum them each time, i assume? the LINQ  `Sum` is 'expensive'.

Not sure you could, as the modifiers can change between shots. The particular use-case I have will be applying an inaccuracy debuff to all units within a radius.

>     * both apply functions might be faster if the function argument was an array and not an enumerable. not sure. perhaps worth testing.

Possibly yea, I don't really know much about C# performance best practices, but if iterating instead of using Sum() then would it avoid this issue?